swarm/storage/hasherstore.go

```diff
@@ -0,0 +1,223 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+	"github.com/ethereum/go-ethereum/swarm/storage/encryption"
+)
+
+type chunkEncryption struct {
+	spanEncryption encryption.Encryption
+	dataEncryption encryption.Encryption
+}
+
+type hasherStore struct {
+	store           ChunkStore
+	hashFunc        SwarmHasher
+	chunkEncryption *chunkEncryption
+	refSize         int64
+	wg              *sync.WaitGroup
+	closed          chan struct{}
+}
+
+func newChunkEncryption() *chunkEncryption {
+	return &chunkEncryption{
+		spanEncryption: encryption.New(0, math.MaxUint32, sha3.NewKeccak256),
+		dataEncryption: encryption.New(4096, 0, sha3.NewKeccak256),
+	}
+}
+
+// NewHasherStore creates a hasherStore object, which implements Putter and Getter interfaces.
+// With the HasherStore you can put and get chunk data (which is just []byte) into a ChunkStore
+// and the hasherStore will take core of encryption/decryption of data if necessary
+func NewHasherStore(chunkStore ChunkStore, hashFunc SwarmHasher, toEncrypt bool) *hasherStore {
+	var chunkEncryption *chunkEncryption
+
+	refSize := int64(hashFunc().Size())
+	if toEncrypt {
+		chunkEncryption = newChunkEncryption()
+		refSize += encryption.KeyLength
+	}
+
+	return &hasherStore{
+		store:           chunkStore,
+		hashFunc:        hashFunc,
+		chunkEncryption: chunkEncryption,
+		refSize:         refSize,
+		wg:              &sync.WaitGroup{},
+		closed:          make(chan struct{}),
+	}
+}
+
+// Puts the chunkData into the ChunkStore of the hasherStore and returns the reference.
+// If hasherStore has a chunkEncryption object, the data will be encrypted.
+// Asynchronous function, the data will not necessarily be stored when it returns.
+func (h *hasherStore) Put(chunkData ChunkData) (Reference, error) {
+	c := chunkData
+	size := chunkData.Size()
+	var encryptionKey encryption.Key
+	if h.chunkEncryption != nil {
+		var err error
+		c, encryptionKey, err = h.encryptChunkData(chunkData)
+		if err != nil {
+			return nil, err
+		}
+	}
+	chunk := h.createChunk(c, size)
+
+	h.storeChunk(chunk)
+
+	return Reference(append(chunk.Key, encryptionKey...)), nil
+}
+
+// Returns data of the chunk with the given reference (retrieved from the ChunkStore of hasherStore).
+// If the data is encrypted and the reference contains an encryption key, it will be decrypted before
+// return.
+func (h *hasherStore) Get(ref Reference) (ChunkData, error) {
+	if h.store == nil {
```

Comment:
Why are we checking for `h.store == nil`? Does this type make sense if h.store is nil?

Reply:
There is one place where we use it with nil, because we only need it for hashing: https://github.com/ethersphere/go-ethereum/blob/swarm-network-rewrite-encryption/cmd/swarm/hash.go#L41
This cmd line tool just gets the hash without storing data. So I think it makes sense to have a nil store if you only need the "hasher" functionality, but not the "store".

Reply:
I think we should just implement a mock/dummy memory store and use that, when we don't care about the storing functionality, rather than add the load of having to check for nil everywhere we use the h.store.

Reply:
Another point is that we keep on passing multiple nil and boolean arguments to methods, which gets very confusing (I think @holisticode has a similar comment in another PR at the moment).

I think we should be encapsulate behaviour with interfaces, rather than in this way.

Reply:
@nonsense Ok, I agree, I implemented this: c03ae51e67eb88ab7f43c3594a0263b0d8e73915
Is this what you meant?

Reply:
Yes, that's one way of doing it, and it is already much better than before. I would have just used the `MemStore`, so that we don't have yet another impl. - we would just push the chunks to a memory store for the duration of the test, which is fine, even if we don't need them.

```
+// Gut doesn't store anything it is just here to implement ChunkStore
+func (d *FakeChunkStore) Get(Key) (*Chunk, error) {
+	return nil, nil
+}